auth0/src/main/java/com/auth0/android/provider/CustomTabsController.java

```diff
@@ -142,9 +142,9 @@ public void run() {
                 intent.setData(uri);
                 try {
                     context.startActivity(intent);
-                } catch (ActivityNotFoundException ignored) {
-                    Intent fallbackIntent = new Intent(Intent.ACTION_VIEW, uri);
-                    context.startActivity(fallbackIntent);
+                } catch (ActivityNotFoundException ex) {
+                    Log.e(TAG, "No Browser available to open the URI");
+                    throw new IllegalStateException("Could not find any Browser application installed in this device to handle the intent.", ex);
```

Comment:
@lbalmaceda I would prefer that you Log.e() this rather than throw the IllegalStateException. We would prefer that over any possibility of a crash. This method, launchUri(), is called from your AuthenticationActivity. As such, it's impossible to catch the exception (without using an UncaughtExceptionHandler).

Reply:
@billgockeler The only public facing class is `WebAuthProvider`; then `OAuthManager` and `CustomTabsController` are package-private and used internally by this lib. This exception is thrown as part of the Controller logic if no Browser app was installed at the moment the Uri was meant to be launched. However, a previous check at `WebAuthProvider.Builder#start` asserts a browser is available or calls the callback with the exception, this happens [here](https://github.com/auth0/Auth0.Android/pull/169/files#diff-163793ac3b8cd48982e0126365834c3eR296). So there's no way this library's WebAuth could be used without that check being performed first. The `AuthenticationActivity` is not meant to be called directly by you.

That said, I don't agree this exception should be caught silently since as I said originally this is an edge case that it doesn't make sense to support. e.g. if there's no browser, the authentication can't be done at all. However, this case is already handled for you and you'll receive the `onFailure` call if that happens.

Reply:
@lbalmaceda I fully understand that we're not calling `AuthenticationActivity` directly. But I think perhaps you're not understanding my point so let me try this another way.

1. What means do we have to catch the `IllegalStateException` that might be thrown by `CustomTabsController`? The answer is none (without installing an UncaughtExceptionHandler). If thrown, it will result in a crash in our app, just as we've been dealing with for months. This is not acceptable.

2. I understand your code and that the `IllegalStateException` should never be thrown. Then why throw it?

Reply:
@lbalmaceda I fully understand that we're not calling `AuthenticationActivity` directly. But I think perhaps you're not understanding my point so let me try this another way.

1. What options do we have to catch the `IllegalStateException` that might be thrown by `CustomTabsController`? The answer is none (without installing an UncaughtExceptionHandler). If thrown, it will result in a crash in our app, just as we've been dealing with for months. This is not acceptable.

2. I understand your code and that the `IllegalStateException` should never be thrown. If so, then why throw it?

Reply:
I'd like this to fail fast, as soon as we can check if a browser is present. We do that in the builder when the _start_ method is called. We can fail safely since the callback is on scope. Not doing it early would mean a lot of instances are getting created (WebAuthProvider, OAuthManager, CustomTabsController, the activity, etc) when we know in advance it will fail, so it's a waste of resources IMO. e.g. we could capture this exception on the AuthenticationActivity and notify somehow the WebAuthProvider that it couldn't even be launched because no browser is installed... Waste of resources 🌮 

I thought of keeping it as part of this "internal class" logic, since the method is "failing" and someone should be notified about that. I do agree that throwing an exception deliberately without catching it is not correct, and I'm not going to catch it there because of what I explained above. Since we both agree it's unreachable code, as I'm checking for browser apps ASAP and failing safely, the 2 options left are either keep or remove this uncaught exception. I've removed it already ✌️ Thanks for the comments